### PR TITLE
Feature/Threshold

### LINKED ISSUE #550: Binary (logic gates) Conditions [Threshold conditions & Inverter Condition]
### Problem 
A SEA condition is designed in a modular way in order to be wired with other conditions in service execution agreement (SEA) but SEA assumes that all the conditions will be executed simultaneously and in one direction without any kind of intermediary checks. The problem is by adding new conditions, we should separate the check points logic from the business logic of the condition itself and have a generic check points as intermediary conditions.

An example of this, the whitelisting template #549  includes four conditions: `lock` and `release` reward, `secret store`, and `whitelisting conditions`. In order to glue them, we need to specify which ones should be executed first, so the typical scenario is calling lock reward, then check the consumer's address is whether whitelisted or not, fulfil the secret store and finally release/refund the reward. 

You might wondering what could happen if somebody fulfils the secret store (based on the current implementation)  before whitelisting condition, this typical scenario won't protect bad actor from abusing the template execution. Therefore, we need a `circuit breaker` in which protects the fulfilment execution of secret store until the reward has been `locked` `AND` whitelisting condition is `fulfilled` without changing the business logic of secret store because adding check point inside the secret store condition logic won't work for all the use cases that we have.

### Introduction
A Logic gate condition acts as a filter for a set of input condition(s) in which sends a signal whether to complete the flow execution or abort it. These conditions work as intermediary conditions where they wire SEA conditions in a way that can support more complex scenarios. The logic gate conditions can be categorized into three basic condition types but also we can expend this to support more features in the future.

image

The following section describes more implementation details about the input parameters for the logic gate condition:

## Condition Types

Each input should be distinguished by a logic gate type (`OR`,`AND` or `INV`) where:
 
**OR**:
Also known as `Thresholding condition`
- `1` of `m` where m is the input conditions and (1 < m)
- `n` of `m` where m is the input conditions and (n < m)

**AND**
Another different type of the `Thresholding condition`
- AND feature (`m` of `m`)  where m is the input conditions and (m > 1)

**INV**
A logic gate inverter which controls the flow as follows:
- It has only one input condition
- { `Fulfilled` --> `Aborted` || `Aborted` --> `Fulfilled` } 

The below flow shows an example for using `Threshold condition`: 

## Input-condition(s)

This array should at least hold one or more conditions Identifiers. Each identifier refers to the condition ID.  For `AND` and `OR` condition types should have at least two input conditions, however `inverter` condition type has only one input condition.

For `AND` logic gate condition type, It assumes or expects only `Fulfilled` input conditions, which means that in case of expecting Aborted condition as an `input` we can mix this by using `inverter` condition type in the input condition then wire the output as `Fulfilled`. The below example describe this:

image

Fortunately this scenario is neither needed in `OR` nor `inverter` logic condition condition type.

## Implementation Details

First we start with the base contract which is the `Logic Gate condition`, a sample implementation details is shown below:
```javascript
pragma solidity 0.5.6;

import './Condition.sol';

contract LogicGateCondition is Condition {
    function initialize(
        address _owner,
        address _conditionStoreManagerAddress
    )
        external
        initializer()
    {
        require(
            _conditionStoreManagerAddress != address(0),
            'Invalid address'
        );
        Ownable.initialize(_owner);
        conditionStoreManager = ConditionStoreManager(
            _conditionStoreManagerAddress
        );
    }

    function hashValues(bytes32[] memory inputConditions, uint256 nInputFulfilledConditions)
        public
        pure
        returns (bytes32)
    {
        return keccak256(abi.encodePacked(inputConditions, nInputFulfilledConditions));
    }
   function _fulfill(
        bytes32 _agreementId,
        bytes32[] memory _inputConditions,
        uint256 n
    )
        internal
        returns (ConditionStoreLibrary.ConditionState)
    {
        return super.fulfill(generateId(_agreementId, hashValues(_inputConditions, n)));
    }
}
```

- `AND` Condition
```javascript
contract ANDCondition is LogicGateCondition {
    function fulfill(
        bytes32 _agreementId,
        bytes32[] memory _inputConditions
    )
        external
        returns (ConditionStoreLibrary.ConditionState)
    {
        require(
              _inputConditions.length >= 2,
              'AND condition requires at least two input conditions'
        );
       bool canFulfill = true;
       ConditionStoreLibrary.ConditionState inputConditionState;
        
       for (uint i=0; i < _inputConditions.length; i++)
       { 
            (,inputConditionState,,,,,) = conditionStoreManager
            .getCondition(_inputConditions[i]);
            if( inputConditionState != ConditionStoreLibrary.ConditionState.Fulfilled )
               canFulfill = false;
               break;
       }
       if ( canFulfill )
            return super._fulfill(generateId(_agreementId, hashValues(_inputConditions, _inputConditions.length)));
       return ConditionStoreLibrary.ConditionState.Aborted;
    }
}
```
- `OR` Condition 
 ```javascript
contract ORCondition is LogicGateCondition {
    function fulfill(
        bytes32 _agreementId,
        bytes32[] memory _inputConditions,
        uint256 n
    )
        external
        returns (ConditionStoreLibrary.ConditionState)
    {
        require(
              _inputConditions.length >= 2 &&
              n <= _inputConditions.length,
              'OR condition requires at least two input conditions'
        );
       bool canFulfill = false;
       uint256 counter = 0;
       ConditionStoreLibrary.ConditionState inputConditionState;
        
       for (uint i=0; i < _inputConditions.length; i++)
       { 
            (,inputConditionState,,,,,) = conditionStoreManager
            .getCondition(_inputConditions[i]);
            if( inputConditionState == ConditionStoreLibrary.ConditionState.Fulfilled )
               counter ++;
            if (counter >= n)
                  canFulfill = true;
                  break;
       }
       if ( canFulfill )
            return super._fulfill(generateId(_agreementId, hashValues(_inputConditions, n)));
       return ConditionStoreLibrary.ConditionState.Aborted;
    }
}
```
- `INV` Condition 
 ```javascript
contract INVCondition is LogicGateCondition {
    function fulfill(
        bytes32 _agreementId,
        bytes32[] memory _inputConditions,
    )
        external
        returns (ConditionStoreLibrary.ConditionState)
    {
        require(
              _inputConditions.length == 1,
              'INV condition requires only one input condition'
        );
       ConditionStoreLibrary.ConditionState inputConditionState;
       (,inputConditionState,,,,,) = conditionStoreManager
            .getCondition(_inputConditions[0]);
       if( inputConditionState == ConditionStoreLibrary.ConditionState.Aborted )
            return super._fulfill(generateId(_agreementId, hashValues(_inputConditions, 1)));
       return ConditionStoreLibrary.ConditionState.Aborted;
    }
}
```

## Templates

Building a circuit can easily verify that you are wiring a template in the right way by using logic gate conditions. It needs only to specify the input condition IDs, by following the same pattern that you have it on the paper and describe it as it is the DID document. A sample of a template is shown below:

image

Lets call this template `SampleTemplate.sol`. The figure shows a sample template in which takes different input conditions, digest them using the intermediary conditions (logic gate conditions) and convert them into a binary output. This pattern will simplify the business logic of the actual conditions and keep them self-contained, decoupled, and totally agnostic conditions.

### Implementation 

- Base template 

```javascript
pragma solidity 0.5.6;

import '../registry/DIDRegistry.sol';
import './AgreementTemplate.sol';

contract BaseTemplate is AgreementTemplate, Common {
    address [] storage conditionTypes;
    AgreementData internal agreementData;
    event AgreementCreated(
        bytes32 indexed _agreementId,
        bytes32 _did,
        address indexed _accessConsumer,
        address indexed _accessProvider,
        uint[]  _timeLocks,
        uint[]  _timeOuts
    );

    struct AgreementDataModel {
        address accessConsumer;
        address accessProvider;
    }

    struct AgreementData {
        mapping(bytes32 => AgreementDataModel) agreementDataItems;
        bytes32[] agreementIds;
    }
    function initialize(
        address _owner,
        address _agreementStoreManagerAddress,
        address _didRegistryAddress,
        address[] memory _conditionTypes
    )
        external
        initializer()
    {
        require(
            _owner != address(0) &&
            _agreementStoreManagerAddress != address(0) &&
            _didRegistryAddress != address(0)
            'Invalid address'
        );

        for ( uint i=0; i < _conditionTypes.length; i++)
        {
              require(
                   _conditionTypes[i] != address(0),
                   'Invalid condition type'
              );
              require(
                   isContract(_conditionTypes[i]),
                   'invalid contract address'
              );
        }
        Ownable.initialize(_owner);

        agreementStoreManager = AgreementStoreManager(
            _agreementStoreManagerAddress
        );

        didRegistry = DIDRegistry(
            _didRegistryAddress
        );        
    }

    function createAgreement(
        bytes32 _id,
        bytes32 _did,
        bytes32[] memory _conditionIds,
        uint[] memory _timeLocks,
        uint[] memory _timeOuts,
        address _accessConsumer
    )
        public
        returns (uint size)
    {
        super.createAgreement(
            _id,
            _did,
            _conditionIds,
            _timeLocks,
            _timeOuts
        );

        address owner = address(0);
        address[] memory providers;
        (owner, , , , providers) = didRegistry.getDIDRegister(_did);

        // storing some additional information for the template
        agreementData.agreementDataItems[_id]
            .accessConsumer = _accessConsumer;

        if (providers.length > 0) {
            agreementData.agreementDataItems[_id]
                .accessProvider = providers[0];
        } else {
            agreementData.agreementDataItems[_id]
                .accessProvider = owner;
        }

        agreementData.agreementIds.push(_id);

        emit AgreementCreated(
            _id,
            _did,
            agreementData.agreementDataItems[_id].accessConsumer,
            agreementData.agreementDataItems[_id].accessProvider,
            _timeLocks,
            _timeOuts
        );

        return agreementData.agreementIds.length;
    }

    function getAgreementData(bytes32 _id)
        external
        view
        returns (
            address accessConsumer,
            address accessProvider
        )
    {
        accessConsumer = agreementData.agreementDataItems[_id].accessConsumer;
        accessProvider = agreementData.agreementDataItems[_id].accessProvider;
    }
}
```

-  SampleTemplate.sol

This template as shown above, has 4 input conditions: `lockReward`, `isWhitelisted`, `isMerkelLeaf` and `isSignedText`. These conditions are connected to `AND` and `OR` intermediary conditions where they are connected to `releaseReward` condition using `AND` condition. The following sample code describes the implementation details for the template.

```javascript
pragma solidity 0.5.6;

import '../conditions/LockRewardCondition.sol';
import '../conditions/rewards/EscrowReward.sol';
import '../conditions/logicgates/ANDCondition.sol';
import '../conditions/logicgates/ORCondition.sol';
import '../conditions/AccessSecretStoreCondition.sol';

contract SampleTemplate is BaseTemplate {
    ANDCondition internal andCondition;
    ORCondition internal orCondition;
    AccessSecretStoreCondition internal accessSecretStoreCondition;
    LockRewardCondition internal lockRewardCondition;
    EscrowReward internal escrowReward;

     function initialize(
        address _owner,
        address _agreementStoreManagerAddress,
        address _didRegistryAddress,
        address _accessSecretStoreConditionAddress,
        address _lockRewardConditionAddress,
        address _escrowRewardAddress,
        address _andConditionAddress,
        address _orConditionAddress
    ){
        // build memory array using condition types
       // TODO: initialize BaseTemplate

       // TODO: initialize SampleTemplate by creating 
       // new instances of the template conditions
       // push to condition types 
       // (hint: in this example, we are going to push AND instance twice)
    }
}
```

### PR DESCRIPTION
**Pull request by 0x3bfc**
Closes #550 

- [x] Condition implementation
- [x] Unit testing (coverage 100 %)
- [x] Documentation

Template will be added in a different PR :)

### CODE REVIEW

File: contracts/conditions/ThresholdCondition.sol [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    *       fulfilled or not.
+    * @param _agreementId agreement identifier
+    * @param _inputConditions array of input conditions IDs
+    * @param nInputFulfilledConditions the required number of fulfilled input conditions
+    * @return condition state (Fulfilled/Aborted)
+    */
+    function fulfill(
+        bytes32 _agreementId,
+        bytes32[] calldata _inputConditions,
+        uint256 nInputFulfilledConditions
```
Comment (diminator) [inline contracts/conditions/ThresholdCondition.sol]: you mean `threshold`?

File: contracts/conditions/ThresholdCondition.sol [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        view
+        returns(bool _fulfill)
+    {
+        uint256 counter = 0;
+        _fulfill = false;
+        ConditionStoreLibrary.ConditionState inputConditionState;
+        ConditionStoreLibrary.ConditionState Fulfilled;
+        Fulfilled = ConditionStoreLibrary.ConditionState.Fulfilled;
+        
+        for (uint i=0; i < _inputConditions.length; i++)
```
Comment (diminator) [inline contracts/conditions/ThresholdCondition.sol]: what's the "infeasible array length" with respect to gas?

File: contracts/conditions/ThresholdCondition.sol (hunk 1 above)
Comment (0x3bfc) [reply contracts/conditions/ThresholdCondition.sol]: yes 👍

File: contracts/conditions/ThresholdCondition.sol (hunk 2 above)
Comment (0x3bfc) [reply contracts/conditions/ThresholdCondition.sol]: good point 👍

File: test/unit/conditions/ThresholdCondition.Test.js [hunk 3]
```diff
@@ -0,0 +1,500 @@
...
+    })
+
+    describe('UNINITIALIZED OR ABORTED input conditions', () => {
+        it('should fail if input conditions are Uninitialized', async () => {
+            const {
+                thresholdCondition,
+                conditionStoreManager,
+                inputConditions,
+                createRole
+            } = await setupTest({ fulfillInputConditions: false })
```
Comment (diminator) [inline test/unit/conditions/ThresholdCondition.Test.js]: doesnt this mean the inputconditions are `UNFULFILLED` but NOT `UNINITIALIZED`?

File: test/unit/conditions/ThresholdCondition.Test.js [hunk 4]
```diff
@@ -0,0 +1,500 @@
...
+
+            await assert.isRejected(
+                thresholdCondition.methods['fulfill(bytes32,bytes32[],uint256)'](
+                    agreementId,
+                    inputConditions,
+                    inputConditions.length,
+                    {
+                        from: createRole
+                    }
+                )
```
Comment (diminator) [inline test/unit/conditions/ThresholdCondition.Test.js]: what's the error message?

File: test/unit/conditions/ThresholdCondition.Test.js (hunk 3 above)
Comment (0x3bfc) [reply test/unit/conditions/ThresholdCondition.Test.js]: you are right will fix this one

File: contracts/conditions/ThresholdCondition.sol (hunk 2 above)
Comment (0x3bfc) [reply contracts/conditions/ThresholdCondition.sol]: load test was added it uses over 1000 conditions.

File: test/unit/conditions/ThresholdCondition.Test.js (hunk 4 above)
Comment (0x3bfc) [reply test/unit/conditions/ThresholdCondition.Test.js]: fixed

**Pull request merged**